packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts

```diff
@@ -0,0 +1,563 @@
+    }
+
+    // Add all instance definitions
+    if (props.instanceDefinitions.length > 20) {
+      throw new Error(`No more than 20 instance definitions are allowed per game server group, given ${props.instanceDefinitions.length}`);
+    }
+    (props.instanceDefinitions || []).forEach(this.addInternalInstanceDefinition.bind(this));
+
+    this.role = props.role ?? new iam.Role(this, 'ServiceRole', {
+      assumedBy: new iam.CompositePrincipal(
+        new iam.ServicePrincipal('gamelift.amazonaws.com'),
+        new iam.ServicePrincipal('autoscaling.amazonaws.com')),
+      managedPolicies: [
+        iam.ManagedPolicy.fromAwsManagedPolicyName('GameLiftGameServerGroupPolicy'),
+      ],
+    });
+    this.grantPrincipal = this.role;
+
+    const resource = new CfnGameServerGroup(this, 'Resource', {
+      gameServerGroupName: this.physicalName,
+      autoScalingPolicy: this.parseAutoScalingPolicy(props),
+      deleteOption: props.deleteOption,
+      balancingStrategy: props.balancingStrategy,
+      gameServerProtectionPolicy: props.protectGameServer ? 'FULL_PROTECTION' : 'NO_PROTECTION',
+      instanceDefinitions: cdk.Lazy.any({ produce: () => this.parseInstanceDefinitions() }),
+      launchTemplate: this.parseLaunchTemplate(props),
+      minSize: props.minSize,
+      maxSize: props.maxSize,
+      roleArn: this.role.roleArn,
+      vpcSubnets: subnetIds,
+    });
+
+    this.gameServerGroupName = this.getResourceNameAttribute(resource.ref);
+    this.gameServerGroupArn = this.getResourceArnAttribute(resource.attrGameServerGroupArn, {
+      service: 'gamelift',
+      resource: 'gameservergroup',
+      resourceName: this.physicalName,
+      arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
+    });
+    this.autoScalingGroupArn = this.getResourceArnAttribute(resource.attrAutoScalingGroupArn, {
+      service: 'autoscaling',
+      resource: 'autoScalingGroup',
+      resourceName: this.physicalName,
+      arnFormat: cdk.ArnFormat.COLON_RESOURCE_NAME,
+    });
+
+  }
+
+  /**
+   * Adds an instance definition that GameLift FleetIQ can use when balancing and automatically scaling instances in the corresponding Auto Scaling group
+   *
+   * @param instanceType An Amazon EC2 instance type designation.
+   * @param weight Instance weighting that indicates how much this instance type contributes to the total capacity of a game server group.
+   */
+  public addInstanceDefinition(instanceType: ec2.InstanceType, weight?: number) {
+    this.addInternalInstanceDefinition({
+      instanceType: instanceType,
+      weight: weight,
+    });
+  }
+
+  /**
+   * Adds an instance definition that GameLift FleetIQ can use when balancing and automatically scaling instances in the corresponding Auto Scaling group
+   *
+   * @param instanceDefinition The instance definition to add
+   */
+  public addInternalInstanceDefinition(instanceDefinition: InstanceDefinition) {
```

Comment:
What's the difference between these two methods? Should this one be private?

Also, what's the use case for this method? In which scenario would it not be enough to provide all instance definitions in the constructor props?

Reply:
Agreed. No real use cases for using this methods outside of the constructor